lib/logstash/inputs/beats.rb

```diff
@@ -99,104 +131,82 @@ def run(output_queue)
+        next if connection.nil? 
+
+        @threadpool.post do
+          handle_new_connection(connection)
         end
       else
-        @logger.warn("Beats input: the pipeline is blocked, temporary refusing new connection.")
+        @logger.warn("Beats input: the pipeline is blocked, temporary refusing new connection.", 
+                     :reconnect_backoff_sleep => RECONNECT_BACKOFF_SLEEP)
         sleep(RECONNECT_BACKOFF_SLEEP)
       end
     end
   end # def run
 
-  public
   def stop
-    # we may have some stuff in the buffer
-    @codec.flush { |event| @output_queue << event }
+    @logger.debug("Beats input: stopping the plugin")
+
     @lumberjack.close rescue nil
-  end
 
-  public
-  def create_event(map, identity_stream, &block)
-    # Filebeats uses the `message` key and LSF `line`
-    target_field = target_field_for_codec ? map.delete(target_field_for_codec) : nil
-
-    if target_field.nil?
-      event = LogStash::Event.new(map)
-      copy_beat_hostname(event)
-      decorate(event)
-      block.call(event)
-    else
-      # All codecs expects to work on string
-      @codec.decode(target_field.to_s, identity_stream) do |decoded|
-        ts = coerce_ts(map.delete("@timestamp"))
-        decoded["@timestamp"] = ts unless ts.nil?
-        map.each { |k, v| decoded[k] = v }
-        copy_beat_hostname(decoded)
-        decorate(decoded)
-        block.call(decoded)
+    # we may have some stuff in the codec buffer
+    transformer = LogStash::Inputs::BeatsSupport::EventTransformCommon.new(self)
+
+    # Go through all the active connection and flush their
+    # codec content, some context data could be lost in this case
+    # but at least the events main data would be persisted.
+    @connections_list.each do |_, v|
+      v.flush do |event|
+        # We might loose some context of the
+        transformer.transform(event)
+        event.tag("beats_input_flushed_by_logtash_shutdown")
+        @output_queue << event
       end
     end
-  end
 
-  # Copies the beat.hostname field into the host field unless
-  # the host field is already defined
-  private
-  def copy_beat_hostname(event)
-    host = event["beat"] ? event["beat"]["hostname"] : nil
-    if host && event["host"].nil?
-      event["host"] = host
-    end
-  end
-
-  private
-  def coerce_ts(ts)
-    return nil if ts.nil?
-    timestamp = LogStash::Timestamp.coerce(ts)
-    return timestamp if timestamp
-
-    @logger.warn("Unrecognized @timestamp value, setting current time to @timestamp",
-      :value => ts.inspect)
-  rescue LogStash::TimestampParserError => e
-    @logger.warn("Error parsing @timestamp string, setting current time to @timestamp",
-      :value => ts.inspect, :exception => e.message)
+    @logger.debug("Beats input: stopped")
   end
 
-  private
-  def invoke(connection, &block)
-    @threadpool.post do
+  # This Method is called inside a new thread
+  def handle_new_connection(connection)
```

Comment:
fixed, not a fan of method level rescue.. probably why I never think about using them